impeller/geometry/geometry_asserts.h

```diff
@@ -16,8 +16,43 @@
 #include "impeller/geometry/vector.h"
 
 inline bool NumberNear(double a, double b) {
-  static const double epsilon = 1e-3;
-  return (a > (b - epsilon)) && (a < (b + epsilon));
+  if (a == b) {
+    return true;
+  }
+  if (std::isnan(a) || std::isnan(b)) {
+    return false;
+  }
+
+  // We used to compare based on an absolute difference of 1e-3 which
+  // would allow up to 10 bits of mantissa difference in a float
+  // (leaving 14 bits of accuracy being tested). Some numbers in the tests
+  // will fail with a bit difference of up to 19 (a little over 4 bits) even
+  // though the numbers print out identically using the float ostream output
+  // at the default output precision. Choosing a max "units of least precision"
+  // of 32 allows up to 5 bits of imprecision.
+  static constexpr float kImpellerTestingMaxULP = 32;
+
+  // We also impose a minimum step size so that cases of comparing numbers
+  // very close to 0.0 don't compute a huge number of ULPs due to the ever
+  // increasing precision near 0. This value is approximately the step size
+  // of numbers going less than 1.0f.
+  static constexpr float kMinimumULPStep = (1.0f / (1 << 24));
+
+  auto adjust_step = [](float v) {
+    if (v > -kMinimumULPStep && v < kMinimumULPStep) {
+      v = v < 0 ? -kMinimumULPStep : kMinimumULPStep;
+    }
+    return v;
```

Comment:
Micro nit:
```suggestion
    return std::copysignf(std::max(std::abs(v), kMinimumULPStep), v);
```
Not sure if I like it better, though. Up to you!

Reply:
Ooh, yes! I missed that function.